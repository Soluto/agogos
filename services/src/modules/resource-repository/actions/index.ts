import applyResourceGroupUpdates from './update';
import applyResourceGroupDeletions from './delete';

export { applyResourceGroupUpdates, applyResourceGroupDeletions };
