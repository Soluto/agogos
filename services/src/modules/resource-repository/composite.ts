import {ResourceRepository, FetchLatestResult, ResourceGroup} from './types';
import {applyResourceGroupUpdates} from './updates';

export class CompositeResourceRepository implements ResourceRepository {
    constructor(protected repositories: ResourceRepository[]) {}

    async fetchLatest(): Promise<FetchLatestResult> {
        const results = await Promise.all(this.repositories.map(repo => repo.fetchLatest()));

        return results.reduce((res1, res2) => ({
            isNew: res1.isNew || res2.isNew,
            resourceGroup: applyResourceGroupUpdates(res1.resourceGroup, res2.resourceGroup),
        }));
    }

    getResourceGroup(): ResourceGroup {
        const rgs = this.repositories.map(r => r.getResourceGroup());

        return rgs.reduce((rg1, rg2) => applyResourceGroupUpdates(rg1, rg2));
    }

    async update(): Promise<void> {
        throw new Error('Multiplexed resource repository cannot handle updates');
    }

    async writePolicyAttachment(): Promise<void> {
        throw new Error('Multiplexed resource repository cannot handle updates');
    }

    public getPolicyAttachment(filename: string): Buffer {
        for (const repo of this.repositories) {
            const policyAttachment = repo.getPolicyAttachment(filename);
            if (policyAttachment) return policyAttachment;
        }

        throw new Error(`Policy attachment with the filename ${filename} was not found`);
    }

    public async initializePolicyAttachments() {
        await Promise.all(this.repositories.map(repo => repo.initializePolicyAttachments()));
    }
}
