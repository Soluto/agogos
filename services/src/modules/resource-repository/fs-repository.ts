import * as envVar from 'env-var';
import { FileSystemStorage } from '../storage';
import { ResourceRepository } from '.';

export class FileSystemResourceRepository extends ResourceRepository {
  protected policyAttachmentsFolderInitialized = false;

  constructor(
    protected storage: FileSystemStorage,
    protected resourceFilePath: string,
    protected policyAttachmentsFolderPath: string,
    protected isRegistry = false
  ) {
    super(storage, resourceFilePath, policyAttachmentsFolderPath, isRegistry);
  }

  async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
    await this.initializePolicyAttachmentsFolder();
    await super.writePolicyAttachment(filename, content);
  }

  async listPolicyAttachments() {
    await this.initializePolicyAttachmentsFolder();
    return super.listPolicyAttachments();
  }

  private async initializePolicyAttachmentsFolder() {
    if (this.policyAttachmentsFolderInitialized) return;

    await this.storage.mkdir(this.policyAttachmentsFolderPath, { recursive: true });
    this.policyAttachmentsFolderInitialized = true;
  }

  static fromEnvironment(options: { isRegistry: boolean } = { isRegistry: false }) {
    const resourceFilePath = envVar.get('FS_RESOURCE_REPOSITORY_PATH').required().asString();
    const policyAttachmentsFolderPath = envVar
      .get('FS_REPOSITORY_POLICY_ATTACHMENTS_FOLDER_PATH')
      .required()
      .asString();

    const fsStorage = new FileSystemStorage();
    return new FileSystemResourceRepository(
      fsStorage,
      resourceFilePath,
      policyAttachmentsFolderPath,
      options.isRegistry
    );
  }
}
