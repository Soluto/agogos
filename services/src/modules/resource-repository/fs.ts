import {ResourceGroup, ResourceRepository} from '.';
import * as envVar from 'env-var';
import {promises as fs} from 'fs';
import * as path from 'path';
import pLimit from 'p-limit';
import * as config from '../config';
import logger from '../logger';
import {FetchLatestResult} from './types';

export class FileSystemResourceRepository implements ResourceRepository {
    protected current?: {mtime: number; rg: ResourceGroup};
    protected policyAttachmentsDirInitialized = false;
    protected policyAttachments: {[filename: string]: Buffer} = {};
    protected policyAttachmentsRefreshedAt?: Date;

    constructor(protected pathToFile: string, protected policyAttachmentsFolderPath: string) {}

    async fetchLatest(): Promise<FetchLatestResult> {
        const stats = await fs.stat(this.pathToFile);

        if (this.current && stats.mtimeMs === this.current.mtime) {
            return {isNew: false, resourceGroup: this.current.rg};
        }

        const contents = await fs.readFile(this.pathToFile, 'utf8');
        const rg = JSON.parse(contents) as ResourceGroup;
        this.current = {mtime: stats.mtimeMs, rg};

        return {isNew: true, resourceGroup: rg};
    }

    async update(rg: ResourceGroup): Promise<void> {
        await fs.writeFile(this.pathToFile, JSON.stringify(rg));
    }

    async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
        await this.initializePolicyAttachmentsDir();

        const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
        await fs.writeFile(filePath, content);
    }

    public getPolicyAttachment(filename: string): Buffer {
        return this.policyAttachments[filename];
    }

    public async initializePolicyAttachments() {
        try {
            await this.refreshPolicyAttachments();
        } catch (err) {
            logger.fatal({err}, 'Failed fetching fs policy attachments on startup');
            throw err;
        }

        setInterval(async () => {
            try {
                await this.refreshPolicyAttachments();
            } catch (err) {
                logger.error(
                    {err},
                    `Failed refreshing fs policy attachments, last successful refresh was at ${this.policyAttachmentsRefreshedAt}`
                );
            }
        }, config.resourceUpdateInterval);
    }

    private async refreshPolicyAttachments() {
        const newRefreshedAt = new Date();

        const allAttachments = await this.getPolicyAttachmentsList();
        const attachmentsToRefresh = allAttachments
            .filter(a => this.shouldRefreshPolicyAttachment(a))
            .map(a => a.filename);

        if (attachmentsToRefresh.length > 0) {
            const newAttachments = await this.getPolicyAttachments(attachmentsToRefresh);
            newAttachments.forEach(a => (this.policyAttachments[a.filename] = a.content));
        }

        this.policyAttachmentsRefreshedAt = newRefreshedAt;
    }

    private shouldRefreshPolicyAttachment({filename, updatedAt}: {filename: string; updatedAt: Date}) {
        if (!this.policyAttachments[filename]) return true;
        if (!this.policyAttachmentsRefreshedAt) return true;

        return updatedAt > this.policyAttachmentsRefreshedAt;
    }

    private async getPolicyAttachmentsList(): Promise<{filename: string; updatedAt: Date}[]> {
        const filenames = await fs.readdir(this.policyAttachmentsFolderPath);
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const stats = await fs.stat(filePath);

                    const updatedAt = stats.mtime;
                    return {filename, updatedAt};
                });
            })
        );
    }

    private async getPolicyAttachments(filenames: string[]): Promise<{filename: string; content: Buffer}[]> {
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const content = await fs.readFile(filePath);

                    return {filename, content};
                });
            })
        );
    }

    private async initializePolicyAttachmentsDir() {
        if (this.policyAttachmentsDirInitialized) return;

        await fs.mkdir(this.policyAttachmentsFolderPath, {recursive: true});
        this.policyAttachmentsDirInitialized = true;
    }

    static fromEnvironment() {
        const pathToFile = envVar
            .get('FS_RESOURCE_REPOSITORY_PATH')
            .required()
            .asString();
        const policyAttachmentsFolderPath = envVar
            .get('FS_REPOSITORY_POLICY_ATTACHMENTS_FOLDER_PATH')
            .required()
            .asString();

        return new FileSystemResourceRepository(pathToFile, policyAttachmentsFolderPath);
    }
}
