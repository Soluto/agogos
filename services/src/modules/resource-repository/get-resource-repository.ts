import * as config from '../config';
import logger from '../logger';
import {
  CompositeResourceRepository,
  FileSystemResourceRepository,
  IResourceRepository,
  ResourceRepository,
  S3ResourceRepository,
} from '.';

export default function getResourceRepository(options?: { isRegistry: boolean }): IResourceRepository {
  const repositories: ResourceRepository[] = [];

  if (config.useS3ResourceRepository) {
    repositories.push(S3ResourceRepository.fromEnvironment(options));
  }

  if (config.useFileSystemResourceRepository) {
    repositories.push(FileSystemResourceRepository.fromEnvironment(options));
  }

  switch (repositories.length) {
    case 0:
      logger.fatal('Must enable at least one resource repository');
      throw new Error('Must enable at least one resource repository');
    case 1:
      return repositories[0];
    default:
      if (options?.isRegistry) {
        logger.fatal('Registry cannot have more than one resource repository');
        throw new Error('Registry cannot have more than one resource repository');
      }
      return new CompositeResourceRepository(repositories);
  }
}
