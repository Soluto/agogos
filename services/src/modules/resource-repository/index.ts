export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
