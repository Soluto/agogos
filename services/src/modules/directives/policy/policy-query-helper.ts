import { gql } from 'apollo-server-core';
import { graphql } from 'graphql';
import { injectArgs } from '../../arguments-injection';
import { RequestContext } from '../../context';
import logger from '../../logger';
import { PolicyQueryVariables, PolicyArgsObject, PolicyDefinition, ResourceMetadata } from '../../resource-repository';
import { PolicyDirectiveExecutionContext, QueryResults } from './types';

export async function getQueryResult(
  ctx: PolicyDirectiveExecutionContext,
  args: PolicyArgsObject = {}
): Promise<QueryResults> {
  const query = ctx.policyDefinition.query;
  if (!query) return;

  const pqLogger = logger.child({ metadata: ctx.policyDefinition.metadata });

  const variables = prepareVariables(args, query.variables);
  const requestContext: RequestContext = { ...ctx.requestContext, ignorePolicies: true };
  const gql = ctx.policyDefinition.query!.gql;
  pqLogger.trace({ variables }, 'Executing policy query...');
  const { data, errors } = await graphql(ctx.info.schema, gql, undefined, requestContext, variables);
  if (errors) {
    pqLogger.error({ errors }, 'Policy query execution failed');
    throw new Error('Policy query execution failed');
  }
  pqLogger.trace({ data }, 'Policy query executed');
  return data;
}

function prepareVariables(
  args: PolicyArgsObject = {},
  variables?: PolicyQueryVariables
): PolicyQueryVariables | undefined {
  if (!variables) return;

  return Object.entries(variables).reduce<PolicyQueryVariables>((policyArgs, [varName, varValue]) => {
    if (typeof varValue === 'string') {
      varValue = injectArgs(varValue, args);
    }
    policyArgs[varName] = varValue;
    return policyArgs;
  }, {});
}

function getPolicyQueryName({ namespace, name }: ResourceMetadata) {
  return `${namespace}___${name}`.replace(/-/g, '_');
}

export function buildPolicyQueryTypeDef(policy: PolicyDefinition) {
  const argStr = policy.args
    ? `(${Object.entries(policy.args)
        .map(([argName, { type }]) => `${argName}: ${type}`)
        .join(',')})`
    : '';

  const policyQueryName = getPolicyQueryName(policy.metadata);

  const policyQueryTypeDef = gql`
    extend type Policy {
        ${policyQueryName}${argStr}: PolicyResult!
          @policyQuery(namespace: "${policy.metadata.namespace}", name: "${policy.metadata.name}")
    }
  `;
  return [policyQueryName, policyQueryTypeDef];
}

declare module '../../context' {
  interface RequestContext {
    /**
     * This flag indicates that request should be resolved without invoking authorization policies evaluation
     */
    ignorePolicies: boolean;
  }
}
