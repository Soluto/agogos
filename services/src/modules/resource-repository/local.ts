import {ResourceGroup, ResourceRepository} from '.';
import {localResourceRepositoryPath} from '../config';
import {promises as fs} from 'fs';

export class LocalResourceRepository implements ResourceRepository {
    protected localPath: string;

    constructor() {
        if (!localResourceRepositoryPath) {
            throw new Error("Must configure 'LOCAL_RESOURCE_REPO_PATH' for Local repository to work!");
        } else {
            this.localPath = localResourceRepositoryPath;
        }
    }

    /** Returns latest resource group */
    fetch(): Promise<ResourceGroup>;
    async fetch(): Promise<ResourceGroup | null> {
        const json = await fs.readFile(this.localPath, 'utf8');
        return JSON.parse(json);
    }

    async update(rg: ResourceGroup): Promise<void> {
        const {etag, ...rgWithoutEtag} = rg;
        await fs.writeFile(this.localPath, JSON.stringify(rgWithoutEtag));
    }
}
