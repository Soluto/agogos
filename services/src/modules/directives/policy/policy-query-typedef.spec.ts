import { print } from 'graphql';
import { PolicyArgsDefinitions, PolicyDefinition, PolicyType } from '../../resource-repository';
import { buildPolicyQueryTypeDef } from './policy-query-helper';

const policies: [string, PolicyArgsDefinitions?][] = [
  ['No args', undefined],
  [
    'Single mandatory arg',
    {
      arg1: {
        type: 'String!',
      },
    },
  ],
  [
    'Single optional arg',
    {
      arg1: {
        type: 'String',
        optional: true,
      },
    },
  ],
  [
    'Single optional arg with default',
    {
      arg1: {
        type: 'String',
        default: '{jwt.claim}',
        optional: true,
      },
    },
  ],
  [
    'Single mandatory arg with default',
    {
      arg1: {
        type: 'String!',
        default: 'hello',
        optional: true,
      },
    },
  ],
  [
    'Multiple argument',
    {
      arg1: {
        type: 'String!',
      },
      arg2: {
        type: 'Int',
        default: '{1}',
        optional: true,
      },
      arg3: {
        type: 'Boolean!',
        default: '{true}',
        optional: true,
      },
    },
  ],
];

describe('Policy query typedef', () => {
  test.each(policies)('%s', (name, args) => {
    const policyDefinition: PolicyDefinition = {
      metadata: {
        namespace: 'policy-query-typedef',
        name: name.toLowerCase().replace(/ /g, '_'),
      },
      type: PolicyType.opa,
      code: 'REGO code here',
      args,
    };

    const [policyQueryName, policyQueryTypeDef] = buildPolicyQueryTypeDef(policyDefinition);
    expect(print(policyQueryTypeDef)).toMatchSnapshot(policyQueryName);
  });
});
