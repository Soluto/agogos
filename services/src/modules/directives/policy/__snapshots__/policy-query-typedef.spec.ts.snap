// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Policy query typedef Multiple argument: policy_query_typedef___multiple_argument 1`] = `
"extend type Policy {
  policy_query_typedef___multiple_argument(arg1: String!, arg2: Int, arg3: Boolean): PolicyResult! @policyQuery(namespace: \\"policy-query-typedef\\", name: \\"multiple_argument\\")
}
"
`;

exports[`Policy query typedef No args: policy_query_typedef___no_args 1`] = `
"extend type Policy {
  policy_query_typedef___no_args: PolicyResult! @policyQuery(namespace: \\"policy-query-typedef\\", name: \\"no_args\\")
}
"
`;

exports[`Policy query typedef Single mandatory arg with default: policy_query_typedef___single_mandatory_arg_with_default 1`] = `
"extend type Policy {
  policy_query_typedef___single_mandatory_arg_with_default(arg1: String): PolicyResult! @policyQuery(namespace: \\"policy-query-typedef\\", name: \\"single_mandatory_arg_with_default\\")
}
"
`;

exports[`Policy query typedef Single mandatory arg: policy_query_typedef___single_mandatory_arg 1`] = `
"extend type Policy {
  policy_query_typedef___single_mandatory_arg(arg1: String!): PolicyResult! @policyQuery(namespace: \\"policy-query-typedef\\", name: \\"single_mandatory_arg\\")
}
"
`;

exports[`Policy query typedef Single optional arg with default: policy_query_typedef___single_optional_arg_with_default 1`] = `
"extend type Policy {
  policy_query_typedef___single_optional_arg_with_default(arg1: String): PolicyResult! @policyQuery(namespace: \\"policy-query-typedef\\", name: \\"single_optional_arg_with_default\\")
}
"
`;

exports[`Policy query typedef Single optional arg: policy_query_typedef___single_optional_arg 1`] = `
"extend type Policy {
  policy_query_typedef___single_optional_arg(arg1: String): PolicyResult! @policyQuery(namespace: \\"policy-query-typedef\\", name: \\"single_optional_arg\\")
}
"
`;
